core/state/access_events.go

```diff
@@ -0,0 +1,314 @@
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+func (ae *AccessEvents) Copy() *AccessEvents {
+	cpy := &AccessEvents{
+		branches:   make(map[branchAccessKey]mode),
+		chunks:     make(map[chunkAccessKey]mode),
+		pointCache: ae.pointCache,
+	}
+	cpy.Merge(ae)
+	return cpy
+}
+
+// AddAccount returns the gas to be charged for each of the currently cold
+// member fields of an account.
+func (ae *AccessEvents) AddAccount(addr []byte, isWrite bool) uint64 {
+	var gas uint64
+	for i := utils.VersionLeafKey; i <= utils.CodeSizeLeafKey; i++ {
+		gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, byte(i), isWrite)
+	}
+	return gas
+}
+
+// MessageCallGas returns the gas to be charged for each of the currently
+// cold member fields of an account, that need to be touched when making a message
+// call to that account.
+func (ae *AccessEvents) MessageCallGas(destination []byte) uint64 {
+	var gas uint64
+	gas += ae.touchAddressAndChargeGas(destination, zeroTreeIndex, utils.VersionLeafKey, false)
+	gas += ae.touchAddressAndChargeGas(destination, zeroTreeIndex, utils.CodeSizeLeafKey, false)
+	return gas
+}
+
+// ValueTransferGas returns the gas to be charged for each of the currently
+// cold balance member fields of the caller and the callee accounts.
+func (ae *AccessEvents) ValueTransferGas(callerAddr, targetAddr []byte) uint64 {
+	var gas uint64
+	gas += ae.touchAddressAndChargeGas(callerAddr, zeroTreeIndex, utils.BalanceLeafKey, true)
+	gas += ae.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, utils.BalanceLeafKey, true)
+	return gas
+}
+
+// ContractCreateInitGas returns the access gas costs for the initialization of
+// a contract creation.
+func (ae *AccessEvents) ContractCreateInitGas(addr []byte, createSendsValue bool) uint64 {
+	var gas uint64
+	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.VersionLeafKey, true)
+	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.NonceLeafKey, true)
+	if createSendsValue {
+		gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BalanceLeafKey, true)
+	}
+	return gas
+}
+
+// AddTxOrigin adds the member fields of the sender account to the witness,
+// so that cold accesses are not charged, since they are covered by the 21000 gas.
+func (ae *AccessEvents) AddTxOrigin(originAddr []byte) {
+	for i := utils.VersionLeafKey; i <= utils.CodeSizeLeafKey; i++ {
+		ae.touchAddressAndChargeGas(originAddr, zeroTreeIndex, byte(i), i == utils.BalanceLeafKey || i == utils.NonceLeafKey)
+	}
+}
+
+// AddTxDestination adds the member fields of the sender account to the witness,
+// so that cold accesses are not charged, since they are covered by the 21000 gas.
+func (ae *AccessEvents) AddTxDestination(targetAddr []byte, sendsValue bool) {
+	for i := utils.VersionLeafKey; i <= utils.CodeSizeLeafKey; i++ {
+		ae.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, byte(i), i == utils.VersionLeafKey && sendsValue)
```

Comment:
`AddTxDestination` will track the VersionKey as write if sendsValue is true.

Any reason for it? It looks a bit confusing.

Reply:
yup, that was a typo, good catch